src/models/StyledComponent.js

```diff
@@ -137,118 +115,80 @@ class BaseStyledComponent extends Component<*, BaseState> {
-      // that by updating when an event is emitted
-    } else if (styledContext !== undefined) {
-      const { subscribe } = styledContext
-      this.unsubscribeId = subscribe(nextTheme => {
-        // This will be called once immediately
-        const theme = determineTheme(
-          this.props,
-          nextTheme,
-          this.constructor.defaultProps
-        )
-
-        const generatedClassName = this.generateAndInjectStyles(
-          theme,
-          this.props
-        )
-
-        this.setState({ theme, generatedClassName })
-      })
-    } else {
-      // eslint-disable-next-line react/prop-types
-      const theme = this.props.theme || EMPTY_OBJECT
-      const generatedClassName = this.generateAndInjectStyles(theme, this.props)
-      this.setState({ theme, generatedClassName })
-    }
-  }
-
-  componentWillReceiveProps(nextProps: { theme?: Theme, [key: string]: any }) {
-    // If this is a statically-styled component, we don't need to listen to
-    // props changes to update styles
-    const { componentStyle } = this.constructor
-    if (componentStyle.isStatic) {
-      return
-    }
-
-    this.setState(prevState => {
-      const theme = determineTheme(
-        nextProps,
-        prevState.theme,
-        this.constructor.defaultProps
-      )
-      const generatedClassName = this.generateAndInjectStyles(theme, nextProps)
-
-      return { theme, generatedClassName }
-    })
-  }
-
-  componentWillUnmount() {
-    this.unsubscribeFromContext()
-  }
-
   render() {
-    // eslint-disable-next-line react/prop-types
     const { innerRef } = this.props
-    const { generatedClassName } = this.state
-    const { styledComponentId, target } = this.constructor
-
-    const isTargetTag = isTag(target)
-
-    const className = [
-      // eslint-disable-next-line react/prop-types
-      this.props.className,
+    const {
       styledComponentId,
-      this.attrs.className,
-      generatedClassName,
-    ]
-      .filter(Boolean)
-      .join(' ')
-
-    const baseProps: Object = {
-      ...this.attrs,
-      className,
-    }
-
-    if (isStyledComponent(target)) {
-      baseProps.innerRef = innerRef
-    } else {
-      baseProps.ref = innerRef
-    }
-
-    const propsForElement = baseProps
-    let key
-
-    for (key in this.props) {
-      // Don't pass through non HTML tags through to HTML elements
-      // always omit innerRef
-      if (
-        key !== 'innerRef' &&
-        key !== 'className' &&
-        (!isTargetTag || validAttr(key))
-      ) {
-        propsForElement[key] =
-          key === 'style' && key in this.attrs
-            ? { ...this.attrs[key], ...this.props[key] }
-            : this.props[key]
-      }
-    }
+      target,
+      defaultProps,
+      componentStyle,
+    } = this.constructor
+    const isTargetTag = isTag(target)
 
-    return createElement(target, propsForElement)
+    return (
+      <ThemeConsumer>
+        {({ theme } = {}) => {
```

Comment:
Can you use the empty object from utils/empties?

Reply:
Can we also cache this function somehow instead of recreating it every render?

Reply:
If I use `EMPTY_OBJECT` I get a flow error though, so the best way might create a `type` for the context and not destructure it, right? I've not really worked with flow before so I might need some help on certain bits

Yeah good point, although I don't it's that expensive to recreate it

Reply:
It's more about memory pressure. This is a hot path, so we want it to be as maximally-efficient as possible.

Reply:
Makes sense, will refactor to take this into account!